client.go

```diff
@@ -10,58 +10,97 @@ import (
 	"github.com/gopcua/opcua/uasc"
 )
 
+var DefaultSessionConfig = uasc.NewClientSessionConfig(
+	[]string{"en-US"},
+	&ua.AnonymousIdentityToken{PolicyID: "open62541-anonymous-policy"},
+)
+
 // Client is a high-level client for an OPC/UA server.
 // It establishes a secure channel and a session.
 type Client struct {
-	Addr string
+	// Endpoint is the endpoint URL the client connects to.
+	Endpoint string
+
+	// Config is the configuration for the secure channel.
+	Config *uasc.Config
+
+	// SessionConfig is the configuration for the session.
+	SessionConfig *uasc.SessionConfig
 
-	config  *uasc.Config
 	sechan  *uasc.SecureChannel
 	session *uasc.Session
 }
 
-func NewClient(addr string, cfg *uasc.Config) *Client {
-	return &Client{Addr: addr, config: cfg}
+// NewClient creates a client object, connects it to the given endpoint with a secure channel and
+// a session.
+func NewClient(endpoint string, cfg *uasc.Config, sessionCfg *uasc.SessionConfig) (*Client, error) {
+	c := &Client{Endpoint: endpoint, Config: cfg, SessionConfig: sessionCfg}
+	if err := c.Connect(); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// Connect connects to the server and establishes a secure channel and a session.
+func (c *Client) Connect() error {
+	if err := c.OpenSecureChannel(); err != nil {
+		return err
+	}
+	return c.CreateSession()
 }
 
-// Open connects to the server and establishes a secure channel
-// and a session.
-func (c *Client) Open() error {
+// OpenSecureChannel connects to the server and establishes a secure channel.
+func (c *Client) OpenSecureChannel() error {
+	if c.sechan != nil {
+		return fmt.Errorf("already connected")
+	}
 	ctx := context.Background()
-	conn, err := uacp.Dial(ctx, c.Addr)
+	conn, err := uacp.Dial(ctx, c.Endpoint)
 	if err != nil {
 		return err
 	}
-	sechan := uasc.NewSecureChannel(conn, nil)
+	sechan := uasc.NewSecureChannel(conn, c.Config)
 	if err := sechan.Open(); err != nil {
 		conn.Close()
 		return err
 	}
-	sechan.EndpointURL = c.Addr
+	sechan.EndpointURL = c.Endpoint
 	c.sechan = sechan
+	return nil
+}
 
-	// todo(fs): this should probably be configurable.
-	sessionCfg := uasc.NewClientSessionConfig(
-		[]string{"en-US"},
-		&ua.AnonymousIdentityToken{PolicyID: "open62541-anonymous-policy"},
-	)
-
-	session := uasc.NewSession(sechan, sessionCfg)
+// CreateSession establishes a new session on the secure channel.
+// todo(fs): can there be more than one session on a secure channel?
```

Comment:
Then `Client` and `Session` are different concepts and we need `CreateSession()` on the `Client` in addition to `ActivateSession`.

Maybe like this:

```go
c := &Client{...}
if err := c.Connect(); err != nil {
    panic(err)
}
defer c.Close()

// create new session
s, err := c.CreateSession()
if err != nil {
    panic(err)
}
defer s.Close()

s.Read(ua.NodeID(2258))

// or activate existing session
s2, err := c.ActivateSession(sessionID)
if err != nil {
    panic(err)
}
defer s2.Close()

s2.Write(ua.NodeID(123), 12)
...
```

we could then have a `NewSession` convenience function like 

```go
func NewSession(endpoint string, cfg *Config) (*Session, error) {
    c := &Client{Endpoint:endpoint, Config:cfg}
    if err := c.Connect(); err != nil {
        return nil, err
    }
    return c.CreateSession()
}

Reply:
Also, what happens to the sessions when the underlying secure channel is closed? Do they need to be closed as well, i.e. does the Client need to keep track of the sessions it has created?

And what happens when you create a session with `NewSession` and then close it. Does that close the client as well? This can't be if they are two separate concepts. `NewSession` would have to return both client and session and you would have to close both which is kinda awkward. 

If we want to have `Session` as a separate concept then we need both `Client` and `Session` IMO and you always need to create both as in the example above.